samples/nrf9160/nrf_cloud_mqtt_multi_service/src/application.c

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <logging/log.h>
+#include <modem/location.h>
+#include <net/nrf_cloud.h>
+#include <stdio.h>
+#include <date_time.h>
+#include <cJSON.h>
+#include <math.h>
+
+#include "application.h"
+#include "temperature.h"
+#include "connection.h"
+
+#include "location_tracking.h"
+
+LOG_MODULE_REGISTER(application, CONFIG_MQTT_MULTI_SERVICE_LOG_LEVEL);
+
+/**
+ * @brief Construct a data device message cJSON object with automatically generated timestamp
+ *
+ * @param appid
+ * @return cJSON* - the timestamped data device message object if successful, NULL otherwise
+ */
+static cJSON *create_timestamped_data_message_object(const char *const appid)
+{
+	cJSON *msg_obj = NULL;
+	int64_t timestamp;
+
+	if (date_time_now(&timestamp)) {
+		LOG_ERR("Failed to create timestamp for data message "
+			"with appid %s", log_strdup(appid));
+		return NULL;
+	}
+
+	msg_obj = cJSON_CreateObject();
+	if (cJSON_AddStringToObject(msg_obj, "appId",		appid)		   == NULL ||
```

Comment:
Good point. I will also make sure to note in documentation and in comments some references to the cloud API.

Not sure whether your idea with the helper functions would be beyond scope -- I think it kind of depends on which helper functions you are thinking would be useful to demonstrate. I can't remember why I ruled out the functions in there, maybe I thought they were too abstracted for this demo? I could be talked out of that, if we decide that teaching people how to actually structure device messages is unimportant.

Reply:
I've switched to using the defines in `nrf_cloud_codec.h`

Also, I realize in this moment that I misread you, regarding your second point.

I could quite possibly make use of `nrf_cloud_encode_sensor_data` to eliminate pretty much all of the cJSON code in this sample.

I'm on the fence about whether it's a good idea to do that, though. It would abstract device message generation behind a relatively opaque abstraction layer. It would also require basically recreating what's already written in the sample if a user decides they want to deviate from the types of messages that `nrf_cloud_codec.h` can generate through its helpers.

I think I will default towards leaving manual construction of device messages in, I think it lends more instructive power to the sample without complicating it too much. Especially since the cJSON stuff is nicely compartmentalized into a couple of functions. But if either of you (@plskeggs @jayteemo) feel otherwise, I could be talked into switching to `nrf_cloud_encode_sensor_data`!